go/libraries/doltcore/env/actions/infer_schema.go

```diff
@@ -0,0 +1,419 @@
+		typeToCount := inf.colType[i]
+		hasNegatives := inf.negatives[i]
+		kind, nullable := typeCountsToKind(name, typeToCount, hasNegatives)
+
+		constraints := make([]schema.ColConstraint, 0, 1)
+		if !nullable {
+			constraints = append(constraints, schema.NotNullConstraint{})
+		}
+
+		tag = nextTag(tag, existingCols)
+		thisTag := tag
+		var col *schema.Column
+		if existingCol, ok := existingCols.GetByName(name); ok {
+			if inf.impArgs.KeepTypes {
+				col = &existingCol
+			} else {
+				thisTag = existingCol.Tag
+			}
+		} else {
+			tag++
+		}
+
+		if col == nil {
+			tmp := schema.NewColumn(name, thisTag, kind, partOfPK, constraints...)
+			col = &tmp
+		}
+
+		if col.IsPartOfPK {
+			pkCols = append(pkCols, *col)
+		} else {
+			cols = append(cols, *col)
+		}
+	}
+
+	if len(pkCols) != len(inf.pkColToIdx) {
+		return nil, errors.New("some pk columns were not found")
+	}
+
+	orderedPKCols := make([]schema.Column, len(pkCols))
+	for _, col := range pkCols {
+		idx, ok := inf.pkColToIdx[col.Name]
+
+		if !ok {
+			return nil, errors.New("could not find key column")
+		}
+
+		orderedPKCols[idx] = col
+	}
+
+	colColl, err := schema.NewColCollection(cols...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	pkColColl, err := schema.NewColCollection(orderedPKCols...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return schema.SchemaFromPKAndNonPKCols(pkColColl, colColl)
+}
+
+func nextTag(tag uint64, cols *schema.ColCollection) uint64 {
+	for {
+		_, ok := cols.GetByTag(tag)
+
+		if !ok {
+			return tag
+		}
+
+		tag++
+	}
+}
+
+func typeCountsToKind(name string, typeToCount map[types.NomsKind]int, hasNegatives bool) (types.NomsKind, bool) {
+	var nullable bool
+	kind := types.NullKind
+
+	for t := range typeToCount {
+		if t == types.NullKind {
+			nullable = true
+			continue
+		} else if kind == types.NullKind {
+			kind = t
+		}
+
+		if kind == t {
+			continue
+		}
+
+		switch kind {
+		case types.StringKind:
+			if nullable {
+				return types.StringKind, true
+			}
+
+		case types.UUIDKind:
+			//cli.PrintErrln(color.YellowString("warning: column %s has a mix of uuids and non uuid strings.", name))
+			kind = types.StringKind
+
+		case types.BoolKind:
+			kind = types.StringKind
+
+		case types.IntKind:
+			if t == types.FloatKind {
+				kind = types.FloatKind
+			} else if t == types.UintKind {
+				if !hasNegatives {
+					kind = types.UintKind
+				} else {
+					//cli.PrintErrln(color.YellowString("warning: %s has values larger than a 64 bit signed integer can hold, and negative numbers.  This will be interpreted as a string.", name))
+					kind = types.StringKind
+				}
+			} else {
+				kind = types.StringKind
+			}
+
+		case types.UintKind:
+			if t == types.IntKind {
+				if hasNegatives {
+					//cli.PrintErrln(color.YellowString("warning: %s has values larger than a 64 bit signed integer can hold, and negative numbers.  This will be interpreted as a string.", name))
+					kind = types.StringKind
```

Comment:
Seems like having negatives should only downgrade to an int, not a string

Reply:
if it is a uint to begin with it's because it doesn't fit in a signed int64, so can't downgrade.